builtin/providers/azurerm/resource_arm_app_gateway.go

```diff
@@ -0,0 +1,1667 @@
+		pool := network.ApplicationGatewayBackendAddressPool{
+			Name: &name,
+			ApplicationGatewayBackendAddressPoolPropertiesFormat: &network.ApplicationGatewayBackendAddressPoolPropertiesFormat{
+				BackendAddresses: &backendAddresses,
+			},
+		}
+
+		backendPools = append(backendPools, pool)
+	}
+
+	return &backendPools
+}
+
+func expandAppGatewayBackendHTTPSettings(d *schema.ResourceData, gatewayID string) *[]network.ApplicationGatewayBackendHTTPSettings {
+	configs := d.Get("backend_http_settings").([]interface{})
+	backendSettings := make([]network.ApplicationGatewayBackendHTTPSettings, 0, len(configs))
+
+	for _, configRaw := range configs {
+		data := configRaw.(map[string]interface{})
+
+		name := data["name"].(string)
+		port := int32(data["port"].(int))
+		protocol := data["protocol"].(string)
+		cookieBasedAffinity := data["cookie_based_affinity"].(string)
+		requestTimeout := int32(data["request_timeout"].(int))
+
+		setting := network.ApplicationGatewayBackendHTTPSettings{
+			Name: &name,
+			ApplicationGatewayBackendHTTPSettingsPropertiesFormat: &network.ApplicationGatewayBackendHTTPSettingsPropertiesFormat{
+				Port:                &port,
+				Protocol:            network.ApplicationGatewayProtocol(protocol),
+				CookieBasedAffinity: network.ApplicationGatewayCookieBasedAffinity(cookieBasedAffinity),
+				RequestTimeout:      &requestTimeout,
+			},
+		}
+
+		if data["authentication_certificate"] != nil {
+			authCerts := data["authentication_certificate"].([]interface{})
+			authCertSubResources := make([]network.SubResource, 0, len(authCerts))
+
+			for _, rawAuthCert := range authCerts {
+				authCert := rawAuthCert.(map[string]interface{})
+				authCertID := fmt.Sprintf("%s/authenticationCertificates/%s", gatewayID, authCert["name"])
+				authCertSubResource := network.SubResource{
+					ID: &authCertID,
+				}
+
+				authCertSubResources = append(authCertSubResources, authCertSubResource)
+			}
+
+			setting.ApplicationGatewayBackendHTTPSettingsPropertiesFormat.AuthenticationCertificates = &authCertSubResources
+		}
+
+		probeName := data["probe_name"].(string)
+		if probeName != "" {
+			probeID := fmt.Sprintf("%s/probes/%s", gatewayID, probeName)
+			setting.ApplicationGatewayBackendHTTPSettingsPropertiesFormat.Probe = &network.SubResource{
+				ID: &probeID,
+			}
+		}
+
+		backendSettings = append(backendSettings, setting)
+	}
+
+	return &backendSettings
+}
+
+func expandAppGatewayHTTPListeners(d *schema.ResourceData, gatewayID string) *[]network.ApplicationGatewayHTTPListener {
+	configs := d.Get("http_listener").([]interface{})
+	httpListeners := make([]network.ApplicationGatewayHTTPListener, 0, len(configs))
+
+	for _, configRaw := range configs {
+		data := configRaw.(map[string]interface{})
+
+		name := data["name"].(string)
+		frontendIPConfigName := data["frontend_ip_configuration_name"].(string)
+		frontendIPConfigID := fmt.Sprintf("%s/frontendIPConfigurations/%s", gatewayID, frontendIPConfigName)
```

Comment:
given there's no guarantee that these URL's won't change in the future, would it make more sense to pass in the FrontEnd IP Configuration ID, rather than the name?